client/notices.go

```diff
@@ -0,0 +1,192 @@
+	// Key is the client notice's key. Must be in "domain.com/key" format.
+	Key string
+
+	// RepeatAfter, if provided, allows the notice to repeat after this duration.
+	RepeatAfter time.Duration
+
+	// Data are optional key=value pairs for this occurrence of the notice.
+	Data map[string]string
+}
+
+// Notify records an occurrence of a "client" notice with the specified options,
+// returning the notice ID.
+func (client *Client) Notify(opts *NotifyOptions) (string, error) {
+	var payload = struct {
+		Action      string            `json:"action"`
+		Type        string            `json:"type"`
+		Key         string            `json:"key"`
+		RepeatAfter string            `json:"repeat-after,omitempty"`
+		Data        map[string]string `json:"data,omitempty"`
+	}{
+		Action: "add",
+		Type:   "client",
+		Key:    opts.Key,
+		Data:   opts.Data,
+	}
+	if opts.RepeatAfter != 0 {
+		payload.RepeatAfter = opts.RepeatAfter.String()
+	}
+	var body bytes.Buffer
+	if err := json.NewEncoder(&body).Encode(&payload); err != nil {
+		return "", err
+	}
+
+	result := struct {
+		ID string `json:"id"`
+	}{}
+	_, err := client.doSync("POST", "/v1/notices", nil, nil, &body, &result)
+	if err != nil {
+		return "", err
+	}
+	return result.ID, err
+}
+
+type NoticesOptions struct {
+	// Type, if set, includes only notices of this type.
+	Type NoticeType
+	// Key, if set, includes only notices with this key.
+	Key string
+	// After, if set, includes only notices that were last repeated after this time.
+	After time.Time
+}
+
+// Notice is a notification whose identity is the combination of Type and Key
+// that has occurred Occurrences number of times.
+type Notice struct {
+	ID            string            `json:"id"`
+	Type          NoticeType        `json:"type"`
+	Key           string            `json:"key"`
+	FirstOccurred time.Time         `json:"first-occurred"`
+	LastOccurred  time.Time         `json:"last-occurred"`
+	LastRepeated  time.Time         `json:"last-repeated"`
+	Occurrences   int               `json:"occurrences"`
+	LastData      map[string]string `json:"last-data,omitempty"`
+	RepeatAfter   time.Duration     `json:"repeat-after,omitempty"`
+	ExpireAfter   time.Duration     `json:"expire-after,omitempty"`
+}
+
+type NoticeType string
+
+const (
+	// Recorded whenever a change is updated: when it is first spawned or its
+	// status was updated.
+	NoticeChangeUpdate NoticeType = "change-update"
+
+	// A client notice reported via the Pebble client API or "pebble notify".
+	// The key and data fields are provided by the user. The key must be in
+	// the format "mydomain.io/mykey" to ensure well-namespaced notice keys.
+	NoticeClient NoticeType = "client"
+
+	// Warnings are a subset of notices where the key is a human-readable
+	// warning message.
+	NoticeWarning NoticeType = "warning"
+)
+
+type jsonNotice struct {
+	Notice
+	RepeatAfter string `json:"repeat-after,omitempty"`
+	ExpireAfter string `json:"expire-after,omitempty"`
```

Comment:
This is quite an interesting trick for me. So I take it that the unmarshalling will type match as well as matching the name, and therefore it will select your extended struct member?

Reply:
Yeah, this approach was new to me as well (I stole it from `client/warnings.go`). I don't think it's about the types, it's because the fields in the struct itself (`jsonNotice`) take precedence over / override the same-named fields in the embedded struct.